File: algorithmsjs/linkedlist.js
function LinkedList() {
	this.head = null;
	this.tail = null;
	this.length = 0;
}
LinkedList.prototype.tail = null;
LinkedList.prototype.append = function( node ) {
	if(this.tail) {
		this.tail.next = node;
		this.tail = node;
	} else {
		this.head = this.tail = node;
	}
	this.length++;
}

LinkedList.prototype.reverse = function (){
	var temp = null;
	var prev = null;

	while( this.head ) {
		temp = this.head.next;
		this.head.next = prev;
		prev = this.head;
		this.head = temp;
	}

	this.head = prev;
}

LinkedList.prototype.reverseRecursive = function( head, temp, prev ) {
	temp = this.head.next;
	this.head.next = prev;
	prev = this.head;
	this.head = temp;

	if( this.head ) {
		this.reverseRecursive( this.head, temp, prev )
	} else {
		this.head = prev;
	}
}

LinkedList.prototype.toString = function () {
	var node = this.head;
	var s = "";
	while( node ) {
		var b = node;
		node = b.next;
		s += b.value + "->"
	}
	return s;
}

function ListNode(value, next) {
	this.value = value;
	this.next = next;
}

var letterString = "abcdefg";
var letterArray = letterString.split('');

var linkedList = new LinkedList();
for(var i = 0; i < letterArray.length; i++ ) {
	linkedList.append( new ListNode(letterArray[i], null ) );
}



console.log( linkedList.toString() )
linkedList.reverseRecursive();
console.log("REVERSE!");
console.log( linkedList.toString() )

File: algorithmsJava/LinkedList/src/LinkedList.java

public class LinkedList {
	private class Node {
		public Object	value;
		public Node		next;
		public Node( Object value, Node next ) {
			this.value = value;
			this.next = next;
		}
	}
	
	/***********/
	Node _head;
	Node _tail;
	int _length;
	
	public LinkedList() {
		
	}
	
	public void insert( Object value ) {
		Node node = new Node( value, null );
		if( this._tail != null ) {
			this._tail.next = node;
			this._tail = node;
		} else {
			this._head = this._tail = node;
		}
	}
	
	public void reverse() {
		Node previous = null;
		Node oldNext = null;
		
		while( this._head != null ) {
			oldNext = this._head.next;
			this._head.next = previous;
			previous = this._head;
			this._head = oldNext;
		}
		
		this._head = previous;
	}
	
	public String toString() {
		String s = "";
		
		Node node = this._head;
		while( node != null ) {
			Node b = node;
			s += b.value.toString() + "->";
			node = b.next;
		}
		
		return s;
	}
	
	/*
	 * insert / remove / head / tail
	 */

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		LinkedList list = new LinkedList();
		
		String[] letters = {"a","b","c","d","e","f"};
		for(String letter : letters ) {
			list.insert( letter );
		}
		
		System.out.println(list);
		list.reverse();
//		list.reverseRecursive( null, null);
		System.out.println( list );
	}

}
